crates/ark/src/lsp/completions/sources/common/subset.rs

```diff
@@ -0,0 +1,61 @@
+//
+// subset.rs
+//
+// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
+//
+//
+
+use tree_sitter::Node;
+use tree_sitter::Point;
+
+use crate::lsp::traits::point::PointExt;
+use crate::treesitter::NodeType;
+use crate::treesitter::NodeTypeExt;
+
+pub(crate) fn is_within_subset_delimiters(
+    x: &Point,
+    subset_node: &Node,
+    subset_type: &NodeType,
+) -> bool {
+    let (open, close) = match subset_type {
+        NodeType::Subset => ("[", "]"),
+        NodeType::Subset2 => ("[[", "]]"),
+        _ => std::unreachable!(),
+    };
+
+    let Some(arguments) = subset_node.child_by_field_name("arguments") else {
+        return false;
+    };
+
+    let n_children = arguments.child_count();
+
+    if n_children < 2 {
+        return false;
+    }
+
+    let Some(open_node) = arguments.child(1 - 1) else {
+        return false;
+    };
+    let Some(close_node) = arguments.child(n_children - 1) else {
+        return false;
+    };
+
+    // Ensure open and closing nodes are the right type
+    if !matches!(
+        open_node.node_type(),
+        NodeType::Anonymous(kind) if kind == open
+    ) {
+        return false;
+    }
+    if !matches!(
```

Comment:
Like, is it possible to have a `Subset` where the first child isn't a `NodeType::Anonymous("[")`?

I think it can be a `NodeType::Error` in some super rare cases, but I'm not 100% sure. I could certainly be more specific about logging an error when its an _unexpected_ mismatch, which would probably make the code clearer. Mostly I'm just being extremely defensive.